### PR TITLE
Pauli flow finding algorithm

### PR DESCRIPTION
**Pull request by d1ssk**
Before submitting, please check the following:
- Make sure you have tests for the new code and that test passes (run `tox`)
- format added code by `black -l 120 <filename>`
- If applicable, add a line to the [unreleased] part of CHANGELOG.md, following [keep-a-changelog](https://keepachangelog.com/en/1.0.0/).

Then, please fill in below:

**Context (if applicable):**

**Description of the change:**
Add Pauli flow finding method following the algorithm algorithm described in EPTCS 343, 2021, pp. 50-101.

**Related issue:**
#93

also see that checks (github actions) pass.
If lint check keeps failing, try installing black==22.8.0 as behavior seems to vary across versions.

### CODE REVIEW

**Comment by d1ssk** [comment]
@masa10-f In the flow/gflow checker, isn't it redundant to examine the properties of the flow and its odd neighbor matrix? This is because dependencies are checked in `get_layers_from_flow`, and if there are any issues, the result will become False there. Indeed, even if the part that examines that matrix is commented out, the tests still pass.

**Comment by masa10-f** [comment]
@d1ssk You're right. I didn't give it much thought. Could you delete matrix check in this PR?

**Review by shinich1 — APPROVED** [review]
To me this looks ready to merge. @masa10-f could you also look through this?

File: graphix/gflow.py [hunk 1]
```diff
@@ -848,3 +1482,21 @@ def get_output_from_flow(flow: dict[int, set]) -> set:
         non_inputs |= correction
     outputs = non_inputs - non_outputs
     return outputs
+
+
+def get_pauli_nodes(meas_planes: dict[int, str], meas_angles: dict[int, float]) -> tuple[set[int], set[int], set[int]]:
+    Lx, Ly, Lz = set(), set(), set()
```
Comment (masa10-f) [inline graphix/gflow.py:1488]: Can you add a docstring here?

File: graphix/gflow.py [hunk 2]
```diff
@@ -848,3 +1482,21 @@ def get_output_from_flow(flow: dict[int, set]) -> set:
...
+            Ly |= {node}
+        elif plane == "ZX" and meas_angles[node] == int(meas_angles[node]):
+            Lz |= {node}
+        elif plane == "ZX" and 2 * meas_angles[node] == int(2 * meas_angles[node]):
+            Lx |= {node}
+        elif plane == "YZ" and meas_angles[node] == int(meas_angles[node]):
+            Ly |= {node}
+        elif plane == "YZ" and 2 * meas_angles[node] == int(2 * meas_angles[node]):
+            Lz |= {node}
+    return Lx, Ly, Lz
```
Comment (masa10-f) [inline graphix/gflow.py:1502]: The notation of the planes seem slightly different. We are now following Backens paper[2021]. Please see the [Backens et al(2021)](https://quantum-journal.org/papers/q-2021-03-25-421/) p.8 or the `Plane` object in `pauli.py`.

File: graphix/gflow.py [hunk 3]
```diff
@@ -195,8 +194,6 @@ def gflowaux(
             for i in range(len(x_col)):
                 node = node_order_col[col_permutation.index(i)]
                 g[non_out_node][node] = x_col[i]
-            for i in range(len(kernels)):
-                g[non_out_node][node_order_col[col_permutation.index(len(x_col) + i)]] = kernels[i]
             if meas_planes[non_out_node] in ["ZX", "YZ"]:
                 g[non_out_node][non_out_node] = sp.true
```
Comment (masa10-f) [inline graphix/gflow.py:198]: Altough I personally like rotation-symmetric notation, "ZX" instead of "XZ", MBQC conventionally uses "XZ" and we are following this notation as in `pauli.Plane`. So, could you revert this change?

File: graphix/gflow.py [hunk 4]
```diff
@@ -346,6 +343,522 @@ def flowaux(
...
+    -------
+    f: dict
+        flow function. g[i] is the set of qubits to be corrected for the measurement of qubit i.
+    l_k: dict
+        layers obtained by flow algorithm. l_k[d] is a node set of depth d.
+    """
+    meas_planes = pattern.get_meas_plane()
+    for plane in meas_planes.values():
+        if plane not in ["X", "Y", "XY"]:
+            return None, None
```
Comment (masa10-f) [inline graphix/gflow.py:652]: As far I remember, we don't use "X" or "Y" notations inside the pattern?

File: graphix/gflow.py [hunk 5]
```diff
@@ -346,6 +343,522 @@ def flowaux(
...
+    """Check if the pattern has a valid Pauliflow. If so, return the Pauliflow and layers.
+
+    Parameters
+    ----------
+    pattern: graphix.Pattern object
+        pattern to be based on
+    mode: str(optional)
+        The Pauliflow finding algorithm can yield multiple equivalent solutions. So there are two options
+            - "single": Returrns a single solution
+            - "all": Returns all possible solutions
```
Comment (masa10-f) [inline graphix/gflow.py:752]: Why is this argument necessary for simply retrieving Pauli flow corrections from pattern?

File: tests/test_gflow.py [hunk 6]
```diff
@@ -153,15 +172,71 @@ def generate_test_graphs() -> list[GraphForTest]:
...
+        meas_planes,
+        meas_angles,
+        False,
+        False,
+        True,
+        "graph with no flow and no gflow but pauliflow",
+    )
+    graphs.append(test_graph)
+
     return graphs
```
Comment (masa10-f) [inline tests/test_gflow.py:211]: Can you add another test case including Y and Z measreuements to confirm that our Pauli flow finding algorithm works correctly?

File: tests/test_gflow.py [hunk 7]
```diff
@@ -153,15 +172,71 @@ def generate_test_graphs() -> list[GraphForTest]:
...
 
 
+def get_rand_graph(n_nodes, edge_prob=0.3, seed=None):
+    np.random.seed(seed)
+    graph = nx.Graph()
+    nodes = list(range(n_nodes))
+    graph.add_nodes_from(nodes)
```
Comment (masa10-f) [inline tests/test_gflow.py:218]: What is the purpose of this method? This seems not called anywhere

**Review by masa10-f — COMMENTED** [review]
@d1ssk I reviewed you code and found some points to ask. Could you check them?
I'm sorry but I couldn't check the detail implementation of Pauli flow algorithm due to the lack of my understandings. So, please add another test case covering the all measurement planes. In addition to this, I find that it would be better to refactor the code relating to `MatGF2` for readability. I'll create another issue for that, and I will understand the detail of Pauli flow finding algorithm by the time addressing refactoring.

**Comment by masa10-f** [comment]
@d1ssk 
If the computational complexity is mentioned in the original paper, can you check the performance of our Pauli flow finding method is consistent with it?

File: tests/test_gflow.py (hunk 7 above)
Comment (d1ssk) [reply tests/test_gflow.py:218]: This method is used in latter part of this test (test_rand_graph). The aim is to detect "positive cosistency" between 'find_' methods and 'verify_' methods by using randomly generated graphs. 
Please tell me if there already is other proper function to generate random graphs or to detect false positive.

Comment (masa10-f) [reply tests/test_gflow.py:274]: Sorry, I missed to check the latter part. I understand your intention.

**Comment by masa10-f** [comment]
@d1ssk Thanks for your commitment!  You can merge this branch now.
